kit/ChildSession.cpp

```diff
@@ -3791,4 +3797,218 @@ void ChildSession::loKitCallback(const int type, const std::string& payload)
+                    return;
+                actCmd = (*_tokens)[0];
+                actSubCmd = (*_tokens)[1];
+            }
+            else if (_tokens->equals(0, "mouse"))
+            {
+                actCmd = (*_tokens)[0];
+                if ((*_tokens)[1].starts_with("type="))
+                {
+                    actSubCmd = (*_tokens)[1].substr(5);
+
+                    // If it is a buttonup and we have a saved buttondown, than exchange it to click
+                    if (actSubCmd == "buttonup" && _session->_lastUiCmdLinesLoggedCount > 0
+                        && _session->_lastUiCmdLinesLogged[_session->_lastUiCmdLinesLoggedCount - 1]._subCmd == "buttondown")
+                    {
+                        if (_session->_lastUiCmdLinesLoggedCount == 1)
+                        {
+                            _session->_lastUiCmdLinesLogged[0]._subCmd = "click";
+                            commandHandled = true;
+                        }
+                        else
+                        {
+                            // drop the previous "buttondown", and change the actual to click
+                            _session->_lastUiCmdLinesLoggedCount = 1;
+                            actSubCmd = "click";
+                            // If "buttondown" generated an undo state change, we should save it too.
+                            _session->_lastUiCmdLinesLogged[0]._undoChange += _session->_lastUiCmdLinesLogged[1]._undoChange;
+                        }
+                    }
+                }
+                else
+                {
+                    actSubCmd = (*_tokens)[1];
+                }
+            }
+            else
+            {
+                actCmd = (*_tokens)[0];
+            }
+
+            // Here we are sure we want to log this command sometime...
+            struct timeval tv;
+            gettimeofday(&tv, NULL);
+            std::time_t actTime = tv.tv_sec - _session->_docManager->getLogUiCmd().getKitStartTimeSec();
+            int actTimeUs = tv.tv_usec;
+
+            // We have to check if undo-count-change happened because of it
+            int undoAct = 0;
+            int undoChg = 0;
+            undoAct = atoi(_session->getLOKitDocument()->getCommandValues(".uno:UndoCount"));
+            // If undo count decrease without an undo .uno:Undo, then it is probably a fake (when cap reached)
+            if (_lastUndoCount!=undoAct && (_lastUndoCount<undoAct || actSubCmd == ".uno:Undo"))
+            {
+                if (undoAct - _lastUndoCount > 0)
+                    undoChg = 1;
+                else
+                    undoChg = -1;
+            }
+            if (commandHandled && undoChg != 0)
+            {
+                // Now, possible only if a buttonup become click
+                _session->_lastUiCmdLinesLogged[0]._undoChange += undoChg;
```

Comment:
This is pretty ugly; as a general thought we should not be grouping inside another object's member (at offset [0]) 's private members like this. It looks like a factoring anti-pattern on several levels.

At a very minimum we should have some shared reference for this _session->_lastUIFoo[0] and use it in this method instead of constantly duplicating this de-reference - surely ?

We are ultimately trying to generate less code fewer changes, easier to review code, with higher entropy.

Reply:
thx, and sorry for it, i made some references for this long things, to solve this issue.
I left the array anyway... [1] / [i] is rarely used.. but if later some more complex log change will be needed, then we may will need more of them.

I was tought the compiler optimize them out... but sure that not helps other reviewers / developers...